### ISSUE TITLE
Improve migrations and add support for plugins

### ISSUE THREAD

**Issue by mark-monteiro** [jellyfin/jellyfin#2538]
- [ ] Load migrations via reflection instead of hard-coding them into the migration runner. A timestamp property will need to be added to each migration and the migration runner updated to sort the migrations by timestamp in order to define the order that the migrations are run.
- [ ] Prefix each migration file name with the same timestamp as used in the property. This will keep the migrations sorted in the correct order when displayed in the filesystem, IDEs etc.
- [ ] Update migration runner to also log the migration namespace along side the name when executing migrations. This will be helpful when executing migrations from multiple plugins.
- [ ] Migration running logic should be in `Emby.Server.Implementations` instead of `Jellyfin.Server`
- [x] Instantiate each migration using the dependency injection framework instead of manually. This way each routine can specify its own dependencies independently and it will also not be necessary to pass `ApplicationHost` or `ILogger` instances to the `Run()` method (#2900)

**Comment by sevenrats** [comment]
@mark-monteiro what is the status of this issue?

**Comment by mark-monteiro** [comment]
Have not worked on any of the remaining tasks here and do not have time any more to work on this.

I created this issue when the migration framework was first implemented and had some ideas of how it could be improved. I'm not sure how much demand there actually is for migration support in plugins, I don't think it's something that has really come up.  I would be okay if this was closed, I'll leave it up to the triage team.

**Comment by Bond-009** [comment]
This looks easy enough

**Comment by srcrist** [comment]
@Bond-009 Were you interested in working on this? Should it be moved to "todo"?

**Comment by LezardSC** [comment]
Hi, I'm searching for a not too hard issue to start working on the project. Is this issue still valid? If yes, I would love to give it a try.

**Comment by mark-monteiro** [comment]
@LezardSC I haven't been involved in the project in a while so I'm not 100% sure, but as far as I know this still hasn't been implemented. It might be a good idea to also check in with the [development chat room](https://jellyfin.org/contact) before starting, especially since it's your first issue in the project.

**Comment by AdrianEdelen** [comment]
I think with the recent changes with EF Core and migration behavior and overall database overhaul this maybe isn't needed anymore? One of the features being actively developed is custom database access for plugins, which should allow plugins to have their own database. In my opinion this effectively covers this issues goal.

**Comment by mark-monteiro** [comment]
> I think with the recent changes with EF Core and migration behavior and overall database overhaul this maybe isn't needed anymore? One of the features being actively developed is custom database access for plugins, which should allow plugins to have their own database. In my opinion this effectively covers this issues goal.

This issue isn't related to the database, it's about a separate migration framework for running code-based migrations on the server

**Comment by bingbeann** [comment]
Hi, I'm looking to contribute to the project and stumbled upon this when browsing good-first-issue.

I noticed migrations are already loaded dynamically via reflection (1) and timestamped (2), leaving the below as pending actions right?

- [ ] Update migration runner to also log the migration namespace along side the name when executing migrations. This will be helpful when executing migrations from multiple plugins.
- [ ] Migration running logic should be in Emby.Server.Implementations instead of Jellyfin.Server

**Comment by EliaLeonardi** [comment]
Hi! I noticed this issue is still open and marked as a good first issue. Since it is from 2020, I wanted to double-check if all the listed requirements (reflection for migrations, timestamp renaming, logging namespaces, and moving the runner logic) are still relevant for the current architecture.

If they are, I would like to take this up.

**Comment by JPVenson** [comment]
> Load migrations via reflection instead of hard-coding them into the migration runner. A timestamp property will need to be added to each migration and the migration runner updated to sort the migrations by timestamp in order to define the order that the migrations are run.

Done

> Prefix each migration file name with the same timestamp as used in the property. This will keep the migrations sorted in the correct order when displayed in the filesystem, IDEs etc.

Done same way as the EFCore migrations. Is also on my plate to be grouped by version too.


> Update migration runner to also log the migration namespace along side the name when executing migrations. This will be helpful when executing migrations from multiple plugins.

Migrations have far more extensive logging build in now, including logging to the startupUI.

> Migration running logic should be in Emby.Server.Implementations instead of Jellyfin.Server

Is planned to be moved into its own assembly even.

> Instantiate each migration using the dependency injection framework instead of manually. This way each routine can specify its own dependencies independently and it will also not be necessary to pass ApplicationHost or ILogger instances to the Run() method (https://github.com/jellyfin/jellyfin/pull/2900)

Done and completly overhauled with multiple migration stages.

**Comment by theguymadmax** [comment]
Thanks for anyone who wanted to contribute to this, but most of the work was already done by Venson, see: https://github.com/jellyfin/jellyfin/pull/13950 

And the remaining issues are planned.

**Comment by Mithul** [comment]
Damn, missed the last 3d of comments. I drafted a PR for the remaining work, not sure if its still needed though.
